### ISSUE TITLE
OpenCV::dnn module (C++) Inference with ONNX @ --rect [768x448] inputs

### ISSUE THREAD

**Issue by JustasBart** [ultralytics/yolov5#10104]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hi guys,

I've only recently found out that yolov5 custom models can be trained as a non-square (rectangular) inputs and it got me really excited as that's exactly what I was looking for - for quite some time!

However, while I was able to train my model with:
```
python3 train.py \
--data /home/*/*/custom.yaml \
--weights yolov5s.pt \
--name sdcard \
--imgsz 768 \
--rect \
--single-cls \
--batch -1 \
--epochs 300 \
--cache ram \
--device 0
```
...and was able to export that model as:
```
python3 export.py \
--weights runs/train/sdcard10/weights/best.pt \
--img 768 448 \
--simplify \
--device 0 \
--include onnx
```
**Netron preview of the .onnx**

![image](https://user-images.githubusercontent.com/40023722/200851269-c8ed802e-42c1-44aa-885d-df75f5cd0dc2.png)

...and was even able to load-infer on my OpenCV's (C++) side:
```
<Qt library via QtCreator>
...
net = cv::dnn::readNetFromONNX(modelPath.toStdString());
net.setPreferableBackend(cv::dnn::DNN_BACKEND_CUDA);
net.setPreferableTarget(cv::dnn::DNN_TARGET_CUDA);
...
cv::Mat blob;
cv::dnn::blobFromImage(input, blob, 1.0/255.0, cv::Size(768, 448), cv::Scalar(), true, false);
net.setInput(blob);

std::vector<cv::Mat> outputs;
net.forward(outputs, net.getUnconnectedOutLayersNames());
...
<Based on the official yolov5 examples - tested on many regular (square) models>
```

The results were not so great :/

![image](https://user-images.githubusercontent.com/40023722/200851725-3cb99044-cf9a-4fae-9fe5-f4a2d0df1f9d.png)

Comparing to export that very same model with:
```
python3 export.py \
--weights runs/train/sdcard10/weights/best.pt \
--img 768 768  \
--simplify \
--device 0 \
--include onnx
```

![image](https://user-images.githubusercontent.com/40023722/200852297-08689008-f824-4795-bbac-b75bd81297f3.png)

At first I thought that maybe this would simply be a scaling issue or something simple like that, but I think that there might be more to this as I'd often enough see behaviour such as this:

![image](https://user-images.githubusercontent.com/40023722/200852747-2d129397-5ce6-49cb-8a96-83d351a3c036.png)

This is where it would often enough cling on to either 2 or more detections... Whereas the square one is consistently correct.

One thing that might be relative here is that this dataset is tiny at only **30 images to train and 10 to val**.

However, I would have thought that not squishing and or padding the data would make the result better rather than worse (You can see how on the detection the confidence is lower presumably because it was trained on as --rect).

This is for 1920x1080 raw input images.

### Question
Has anyone attempted/succeeded at doing this?
Would any one have any clues/leads as per what might be worth investigating here in this case?

How is the detect.py -dnn handling this differently from OpenCV's dnn?
Sample of:
```
python3 detect.py --weights runs/train/sdcard10/weights/best.onnx --imgsz 768 448 --device 0 --dnn
```
![07 11 2022_11:15:17:376](https://user-images.githubusercontent.com/40023722/200854249-f9edca6f-f414-4f45-b3cd-59831d190f27.png)

^ This is also what gives me at least some confidence that the actual .onnx file does not appear to be broken (corrupted) as a result of the export.py...

tl;dr: Send some C++ ONNX @--rect model help! :rocket:

**Comment by glenn-jocher** [comment]
@JustasBart as you've shown inference with default code works correctly, therefore the issue is with your inference implementation, which is outside the scope of our support. If you find any bugs or problems in unmodified YOLOv5 code though please let us know :)

**Comment by glenn-jocher** [comment]
@JustasBart also for C++ inference example see Export tutorial at https://docs.ultralytics.com/yolov5/tutorials/model_export

<img width="989" alt="Screenshot 2022-11-09 at 15 26 16" src="https://user-images.githubusercontent.com/26833433/200856023-0827ccc4-1da6-4763-95e3-551f24034835.png">

**Comment by JustasBart** [comment]
@glenn-jocher Thanks for your reply, I'll try and see if I can figure this out on my own though hopefully someone might nudge me in the right direction especially if I get stuck... Once this is solved I propose we should have it as a third example one the repo for everyone to tinker with! :rocket:

**Comment by glenn-jocher** [comment]
@JustasBart oh yes, that's correct. We follow the torch BCHW convention here for --imgsz height width :)

**Comment by Henryplay** [comment]
> @glenn-jocher我找到了解决方案！
> 
> 问题出在 export.py 我正在做的是：
> 
> ```
> python3 export.py \
> --weights runs/train/sdcard10/weights/best.pt \
> --img 768 448 \
> --simplify \
> --device 0 \
> --include onnx
> ```
> 
> 对于宽度：768 和高度：448，
> 
> 但事实证明，正确的做法是：
> 
> ```
> python3 export.py \
> --weights runs/train/sdcard10/weights/best.pt \
> --img 448 768 \
> --simplify \
> --device 0 \
> --include onnx
> ```
> 
> 我没有意识到 export.py 的输入形状是 [H, W]！

Learned

**Comment by glenn-jocher** [comment]
@Henryplay glad to hear you found the solution! If you have any more questions or need further assistance, feel free to ask. Good luck with your YOLOv5 project!
